### PR TITLE
Idea on ProtocolHandler inheritance

### PR DESCRIPTION
**Pull request by ches**
Heya,

I had to try this out myself, so at that point I figured it was easier to share code than write out a lot of abstract babble on the upstream PR 😄 

I did a fair bit of Python in the past, but it's been years now, so I'm brushing up too. This can be some of the weirdest stuff to adjust to because mixins are just a loosely-defined convention in Python and they're still really inheritance at the end of the day, so the lines get rather blurred. And it's multiple inheritance, which not many common languages have, patterns aren't as familiar.

It's a fairly minor thing, but it was bugging me that `EnsimeClient` didn't mix in/inherit from `ProtocolHandler`, because that would express the dependency that does exist and also make `EnsimeClient` more formally abstract: if you try to instantiate one, `NotImplementedError`s _will_ happen for handler methods instead of an `AttributeError` for missing `handle_incoming_response()`.

When I tried to add it, it wouldn't quite work at first because of the way constructors and `super` calls were chaining. This change made it work.

When a class doesn't define `__init__`, it inherits its parent's constructor, but in the context of Python's multiple inheritance, a more apt way to put it might be that _it delegates up any chain of parents_. So like, `EnsimeClientV1(ProtocolHandlerV1, EnsimeClient)` now acts like it has `EnsimeClient`'s constructor signature, but if `ProtocolHandlerV1` were a mixin useful in other contexts, it could be transparent in those chains too. Kinda fits the duck-typey world of Python I guess.

For a mixin class with no state, this is probably a boon for flexibility generally. When state gets involved then explicit constructors have to be dealt with, as is the case when we get down to the `ProtocolHandler` base class. It works out here because `EnsimeClient` explicitly calls `super(EnsimeClient, self).__init__()` and all of its parents also call up the chain already with the same constructor sig, reaching `ProtocolHandler`. This pattern gets referred to as "cooperative multiple inheritance"—[this article](https://rhettinger.wordpress.com/2011/05/26/super-considered-super/) is good, in fact it's linked from the official Python documentation for `super` referencing how it's novel to multiple inheritance.

So this change makes things _less_ explicit in some ways, but I look at it as deferring decisions about how the hierarchy cooperates. I reckon we'll be doing more refactoring in the future, there's still a lot of tight coupling. And it does make `EnsimeClient` truthful about a relationship to `ProtocolHandler` as noted above. 

What do you think?

### CODE REVIEW

File: ensime_shared/ensime.py [hunk 1]
```diff
@@ -822,15 +826,11 @@ def detect_row_column_start():
 
 
 class EnsimeClientV1(ProtocolHandlerV1, EnsimeClient):
-    def __init__(self, *args, **kwargs):
```
Comment (ktonga) [inline ensime_shared/ensime.py:825]: Have you tried it? This was my first approach but I had a "compile" error so I had to add the explicit calls due to the different amount of args in the `init` methods.

**Comment by ktonga** [comment]
I researched a bit about Python inheritance before starting with the refactor, I read about this "Cooperative Mixin" which is kinda the "Cake Pattern" in Scala world. But in order to use it I was forced to add the same `init` args `EnsimClient` has to all `ProtocolHandler`s despite they don't need/use them, just for them to propagate the args during the cooperative initialization chain. I found it pretty horrible.

TBH, I don't like inheritance at all (I hate it), it complicates the modularity, the reasoning and the testability of things a lot without clear gain. I'd much more prefer to use composition here. Do you think we could really isolate the handler from the client so that we can then decide what protocol (instance) the client should use? Maybe changing the signature of the handlers to not depend on client's internal state.

Another component I envision is a sort of a `CommandHandler` which attends user actions and turn them into Ensime requests or something along these lines. This way all the component could truly be tested in isolation and the client would be just the simple glue between them to put command handling, server interaction and response handling to work together. This way maybe, just maybe we could limit the side effects to just one component and make the testing of the other really simple, and make the one-harder-to-test limited in functionality.

Any thoughts? is it possible for you or I'm just dreaming?

Thank you so much for the help, I think this PR is more cooperative than Python's inheritance :stuck_out_tongue_winking_eye:

**Comment by ktonga** [comment]
That's good since we might be using the same Python version you did back in these days :smile: 

> I did a fair bit of Python in the past, but it's been years now,

File: ensime_shared/ensime.py (hunk 1 above)
Comment (ches) [inline ensime_shared/ensime.py:825]: Ugh, so sorry, I was running the code on this branch but I wrote an en-dash in a docstring in the last commit and didn't realize that broke Python at the last minute 😩   Fixed now, this code should run.

**Comment by ches** [comment]
> TBH, I don't like inheritance at all (I hate it), it complicates the modularity, the reasoning and the testability of things a lot without clear gain. I'd much more prefer to use composition here.

The irony is that's what mixins are supposed to be (composition in the `trait` sense at least), but they're still really inheritance 😳 

> Do you think we could really isolate the handler from the client so that we can then decide what protocol (instance) the client should use? Maybe changing the signature of the handlers to not depend on client's internal state.

The method signatures are already good for stateless I think (call ID, payload), the pain is all the calls within their bodies to stuff defined on the client class, and that the handlers are pure procedural side effect. This is definitely what I wanted to decouple going forward—I'm not exactly sure how yet design-wise, but I don't think we can go too far with that for the scope of the protocol v2 split PR right now. I don't think the changes we're making leave things any _worse_ for future refactoring efforts…

> Another component I envision is a sort of a `CommandHandler` which attends user actions and turn them into Ensime requests or something along these lines.

Heh, so I've actually been trying to work on a standalone ENSIME client hoping it could be `pip install ensime` and shareable by ensime-vim and ensime-sublime (and unit tested from the start…). It's not to a state that's ready to publish yet, but this sounds close to one of my big remaining design questions. Would be good to chat more, your ideas may be helpful.

My hope is that it would largely reduce ensime-vim to being _nothing but_ the handlers, almost. I'm modeling all the protocol request and response messages with `namedtuple` which is pretty similar to a Python case class. A client instance provided by the library would send and return these, but I'm trying to come up with what to do for async calls—accept a callback function, or could/should it provide some kind of higher-level API than that? At least initially I think it'll start low-level and leave the rest to implementors. Maybe we'll come up with patterns in ensime-vim that are general enough to upstream as a higher-level API.

I hope I can make progress on it before ensime-vim refactoring discussion runs very far. This protocol version stuff is already good for helping me think about how it should work for calling code like ensime-vim.

Anyway, back to the topic immediately at hand…

**Comment by ktonga** [comment]
> Would be good to chat more, your ideas may be helpful.

Feel free to add me to Hangouts with my gmail address or start a private chat on Gitter if you want. Cheers.

**Pull request merged**